src/runtime/hexagon_remote/halide_remote.cpp

```diff
@@ -44,16 +87,227 @@ int halide_do_task(void *user_context, halide_task_t f, int idx,
+        if (!owned_job) {
+            // If this threads doesn't own a job, it looks for one
+            // and tries to do it. If it cannot find a job, it goes
+            // to sleep.
+            job = work_queue.jobs;
+            if (!job) {
+                // release the lock and go to sleep.
+                // ***************************
+                // *** Work queue unlocked ***
+                // ***************************
+                FARF(LOW, "HVX_TP: %d: goto_work: Couldn't find a job, going to sleep\n", tid);
+                unlock_work_queue();
+                qurt_sem_down(&wait_for_work);
+                FARF(LOW, "HVX_TP: %d: goto_work: Got woken up. Going to look for work.\n", tid);
+                 // ***********************
+                // *** Lock work queue ***
+                // ***********************
+                lock_work_queue();
+                continue;
+            } else {
+                myjob = *job;
+                FARF(LOW, "HVX_TP: %d: goto_work: Found a job working on x = %d\n", tid, myjob.next);
+            }
+        } else {
+            // We are here only if this thread owns a job that is not done.
+            // So do a part of it.
+            if (owned_job->claimed()) {
+                // the owner goes to sleep till the workers wake it up.
+                FARF(LOW, "HVX_TP: %d: goto_work: Owner about to sleep\n", tid);
+                qurt_cond_wait(&owned_job->wakeup_owner, &work_queue.work_mutex);
+                FARF(LOW, "HVX_TP: %d: goto_work: Owner waking up.\n", tid);
+                // This thread should have the lock now after having been woken up.
+                break;
+            } else {
+                myjob = *owned_job;
+                job = owned_job;
+                FARF(LOW, "HVX_TP: %d: goto_work: Owner about to work\n", tid);
+            }
+        }
+        job->next++;
+        job->active_workers++;
+
+        // If all tasks of the job have been claimed, then pop the job off the stack.
+        if (job->claimed()) {
+            work_queue.jobs = job->next_job;
+            FARF(LOW, "HVX_TP: %d: goto_work: All tasks in job claimed. Popping job from stack.\n", tid);
+        }
+        // ***************************
+        // *** Work queue unlocked ***
+        // ***************************
+        unlock_work_queue();
+
+        FARF(LOW, "HVX_TP: %d: goto_work: About to do_task, user_context = 0x%x, f = 0x%x x = %d \n", tid, myjob.user_context, myjob.f, myjob.next);
+        int result = halide_do_task(myjob.user_context, myjob.f,
+                                    myjob.next, myjob.closure);
+        FARF(LOW, "HVX_TP: %d: goto_work: Finished do_task with status = %d\n", tid, result);
+        // ***********************
+        // *** Lock work queue ***
+        // ***********************
+        lock_work_queue();
         if (result) {
-            return result;
+            job->exit_status = result;
+        }
```

Comment:
Hmm, I have thought about this. I want to leave this as is because halide_do_par_for returns the job status and any error reported will report that halide_do_par_for failed, which is cleaner, IMHO.

The only drawback of this is that the other threads will work needlessly, which is not efficient. But again, is efficiency required, when the job is doomed anyway.

If you still feel strongly about this, I'll put a `qurt_thread_exit(result);` line in here.